Regenerate ModuleProgramExecutable unlinkedCodeBlock after being destroyed by low memory cleaning

hyjorc1 opened the pull request:
#### e9cb06ebfbf530f4d571fa5103ebd3d3d6f9aad6
<pre>
Regenerate ModuleProgramExecutable unlinkedCodeBlock after being destroyed by low memory cleaning
<a href="https://bugs.webkit.org/show_bug.cgi?id=255044">https://bugs.webkit.org/show_bug.cgi?id=255044</a>
rdar://104937190

Reviewed by Mark Lam and Yusuke Suzuki.

ModuleProgramExecutable&apos;s unlinkedCodeBlock can be cleared by ScriptExecutable::clearCode
during low memory handling. As a result, ScriptExecutable::newCodeBlockFor can encounter
a null unlinkedCodeBlock. This patch fixes this issue by regenerating the unlinkedCodeBlock
if needed.

No test is added because the manifestation of this bug depends on a race condition.
The race condition requires that ScriptExecutable::clearCode be called after
ModuleProgramExecutable is instantiated and before ScriptExecutable::newCodeBlockFor
runs. This can happen due to low memory handling. We manually simulated this condition
in local testing to reproduce the issue by forcing ScriptExecutable::clearCode to be called
after ModuleProgramExecutable is instantiated and just before the VM returns to the idle loop.
Using this method, we confirmed that the fix works.

* Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
(JSC::ModuleProgramExecutable::generateOrGetUnlinkedCodeBlock):
(JSC::ModuleProgramExecutable::create):
* Source/JavaScriptCore/runtime/ModuleProgramExecutable.h:
* Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
(JSC::ScriptExecutable::newCodeBlockFor):

Canonical link: <a href="https://commits.webkit.org/262788@main">https://commits.webkit.org/262788@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/75e8f0e27d9702a92962919ebb1ed65d82802678

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2600 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2604 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2728 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/4003 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2993 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2745 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2692 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2298 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2624 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/15/builds/3094 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2331 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3769 "Built successfully") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/39/builds/587 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2316 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/2191 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/2176 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2673 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2336 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3544 "Passed tests") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/2494 "Built successfully and passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2367 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/2160 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/2678 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2372 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2319 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/622 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/642 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2355 "Built successfully") | | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/24/builds/2734 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2504 "Built successfully") | | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/3/builds/736 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (changes requested):
Discussed offline. We should make this never nullptr or we should make it work even if it gets nullptr. Throwing an error is semantically incorrect.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,31 +41,37 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
 }
 
-ModuleProgramExecutable* ModuleProgramExecutable::create(JSGlobalObject* globalObject, const SourceCode& source)
+UnlinkedModuleProgramCodeBlock* ModuleProgramExecutable::generateOrGetUnlinkedCodeBlock(JSGlobalObject* globalObject, ThrowScope& throwScope, ParserError& error)
```
The best common practice is to DECLARE_THROW_SCOPE in this function itself instead of passing it in.  There are very few places where it makes sense to pass it in.  This is not one.

Constellation replied:
Do not take `ThrowScope& throwScope`. If the function throws an error, you should define in as `auto scope = DECLARE_THROW_SCOPE(vm)` instead.

Constellation replied:
Do not take `ThrowScope& throwScope`. If the function throws an error, you should define in as `auto scope = DECLARE_THROW_SCOPE(vm)` instead.

Constellation commented on Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
```diff
@@ -268,6 +273,17 @@ CodeBlock* ScriptExecutable::newCodeBlockFor(CodeSpecializationKind kind, JSFunc
+
+        UnlinkedModuleProgramCodeBlock* unlinkedCodeBlock = executable->unlinkedCodeBlock();
+        ParserError error;
+        if (UNLIKELY(!unlinkedCodeBlock))
```
Why is this necessary? If the function is `generateOrGetUnlinkedCodeBlock`, we should always call it since it becomes `get`.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,31 +41,37 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
+    m_unlinkedCodeBlock.set(vm, this, unlinkedModuleProgramCode);
+    VirtualRegister symbolTableReg = VirtualRegister(unlinkedModuleProgramCode->moduleEnvironmentSymbolTableConstantRegisterOffset());
+    SymbolTable* symbolTable = jsCast<SymbolTable*>(unlinkedModuleProgramCode->getConstant(symbolTableReg));
+    m_moduleEnvironmentSymbolTable.set(vm, this, symbolTable->cloneScopePart(globalObject->vm()));
```
Don't need `globalObject->vm()`.  You already have `vm`.

MenloDorian commented on Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
```diff
@@ -268,6 +273,17 @@ CodeBlock* ScriptExecutable::newCodeBlockFor(CodeSpecializationKind kind, JSFunc
+        if (UNLIKELY(!unlinkedCodeBlock)) {
+            throwException(globalObject, throwScope, error.toErrorObject(globalObject, executable->source()));
+            return nullptr;
+        }
```
You can just `RELEASE_ASSERT(unlinkedCodeBlock)` instead in the if block after `generateOrGetUnlinkedCodeBlock` above.

MenloDorian commented on Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
```diff
@@ -295,11 +304,7 @@ CodeBlock* ScriptExecutable::newCodeBlockFor(CodeSpecializationKind kind, JSFunc
-        return nullptr;
-    }
-
+    ASSERT(unlinkedCodeBlock);
```
Let's do this separately as it introduces risks.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.h:
```diff
@@ -52,6 +52,8 @@ class ModuleProgramExecutable final : public GlobalExecutable {
         return bitwise_cast<ModuleProgramCodeBlock*>(Base::codeBlock());
     }
 
+    UnlinkedModuleProgramCodeBlock* getUnlinkedCodeBlock(JSGlobalObject*);
```
nit: I would prefer that you add an enum class here like this:
```
enum class PossibleExceptionsExpected { Yes, No };
```

... and pass that into `getUnlinkedCodeBlock`.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,32 +41,43 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
-        throwVMError(globalObject, scope, error.toErrorObject(globalObject, executable->source()));
+        throwVMError(globalObject, throwScope, error.toErrorObject(globalObject, source()));
         return nullptr;
     }
```
Here inside this if block, `RELEASE_ASSERT(possibleExceptionsExpected == PossibleExceptionsExpected::Yes)`.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,32 +41,45 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
+ModuleProgramExecutable* ModuleProgramExecutable::create(JSGlobalObject* globalObject, const SourceCode& source)
+{
+    VM& vm = globalObject->vm();
+    auto throwScope = DECLARE_THROW_SCOPE(vm);
```
nit: FYI, we don't strictly need this, since this function doesn't itself throw, its children does.  However, having it does not hurt anything.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,32 +41,45 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
 
-    return executable;
+    ModuleProgramExecutable* executable = new (NotNull, allocateCell<ModuleProgramExecutable>(vm)) ModuleProgramExecutable(globalObject, source);
+    executable->finishCreation(globalObject->vm());
```
nit: if you keep the above ThrowScope, then you can use `vm` here.

MenloDorian commented on Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
```diff
@@ -41,32 +41,45 @@ ModuleProgramExecutable::ModuleProgramExecutable(JSGlobalObject* globalObject, c
+    ModuleProgramExecutable* executable = new (NotNull, allocateCell<ModuleProgramExecutable>(vm)) ModuleProgramExecutable(globalObject, source);
+    executable->finishCreation(globalObject->vm());
+    executable->getUnlinkedCodeBlock(globalObject, PossibleExceptionsExpected::Yes); // This generates and binds unlinked code block.
+    RELEASE_AND_RETURN(throwScope, executable);
```
nit: if you remove the above ThrowScope, then make this a plain `return`.

MenloDorian reviewed (approved):
r=me

Constellation reviewed (approved):
r=me

hyjorc1 committed:
Regenerate ModuleProgramExecutable unlinkedCodeBlock after being destroyed by low memory cleaning
https://bugs.webkit.org/show_bug.cgi?id=255044
rdar://104937190

Reviewed by Mark Lam and Yusuke Suzuki.

ModuleProgramExecutable's unlinkedCodeBlock can be cleared by ScriptExecutable::clearCode
during low memory handling. As a result, ScriptExecutable::newCodeBlockFor can encounter
a null unlinkedCodeBlock. This patch fixes this issue by regenerating the unlinkedCodeBlock
if needed.

No test is added because the manifestation of this bug depends on a race condition.
The race condition requires that ScriptExecutable::clearCode be called after
ModuleProgramExecutable is instantiated and before ScriptExecutable::newCodeBlockFor
runs. This can happen due to low memory handling. We manually simulated this condition
in local testing to reproduce the issue by forcing ScriptExecutable::clearCode to be called
after ModuleProgramExecutable is instantiated and just before the VM returns to the idle loop.
Using this method, we confirmed that the fix works.

* Source/JavaScriptCore/runtime/ModuleProgramExecutable.cpp:
(JSC::ModuleProgramExecutable::generateOrGetUnlinkedCodeBlock):
(JSC::ModuleProgramExecutable::create):
* Source/JavaScriptCore/runtime/ModuleProgramExecutable.h:
* Source/JavaScriptCore/runtime/ScriptExecutable.cpp:
(JSC::ScriptExecutable::newCodeBlockFor):

Canonical link: https://commits.webkit.org/262788@main
